File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/properties/SEScalar.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.properties;

import java.lang.reflect.Method;
import java.math.BigDecimal;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.SEEqualOptions;
import mil.tatrc.physiology.datamodel.bind.ScalarData;
import mil.tatrc.physiology.datamodel.exceptions.InvalidUnitException;
import mil.tatrc.physiology.utilities.DoubleUtils;
import mil.tatrc.physiology.utilities.Log;
import mil.tatrc.physiology.utilities.UnitConverter;

/**
 * @author abray
 *
 */
public class SEScalar
{
  protected String             unit;
  protected double             value;
  protected int                precision=-1;
  
  public SEScalar()
  {
    super();
    invalidate();
  }
  
  public SEScalar(double val)
  {
  	super();
    invalidate();
  	unit="";
  	value=val;
  }

  public boolean set(SEScalar s)
  {
    if (s == null || !s.isValid())
      return false;
    this.invalidate();
  
    this.unit = s.unit;
    this.value = s.value;
    this.precision = s.precision;
    
    return true;
  }

  public boolean loadData(Object in) throws ClassCastException
  {
    if (in instanceof ScalarData)
      return load((ScalarData) in);
    else if (in instanceof SEScalar)
      return set((SEScalar) in);
    return false;
  }

  public boolean load(ScalarData in)
  {
    if (in == null)
      return false;
    if(in.getUnit()!=null)
    	this.setValue(in.getValue(),in.getUnit());
    else
    	this.setValue(in.getValue());
    return this.isValid();
  }

  public Object unloadData()
  {
    return unload();
  }

  public ScalarData unload()
  {
    if(!this.isValid())
      return null;
    ScalarData to = CDMSerializer.objFactory.createScalarData();
    unload(to);
    return to;
  }

  protected void unload(ScalarData to)
  {
  	if(!this.unit.equals(""))
  		to.setUnit(this.unit);
  	if(this.precision!=-1)
  	{
  		BigDecimal d = new BigDecimal(this.value);
  		to.setValue(Double.valueOf(d.setScale(3, BigDecimal.ROUND_CEILING).toString()));
  	}
  	else
  		to.setValue(this.value); 
  }

  public boolean equals(Object other)
  {
    if ((other == null) || !(SEScalar.class.isAssignableFrom(this.getClass())))
      return false;
  
    return SEScalar.equals(this, (SEScalar) other, null);
  }
  
  public static boolean equals(SEScalar from, SEScalar to)
  {
    return equals(from, to, null);
  }
  
  public static boolean equals(SEScalar from, SEScalar to, SEEqualOptions options)
  {
    return equals(from, to, options, null);
  }

  public static boolean equals(SEScalar from, SEScalar to, SEEqualOptions options, String name)
  {
    if (options != null)
    {
      if (name == null)
        name = "SEScalar";
      options.pushTrack(name);
    }
    boolean equals=true;
    if(!equals(from.value, from.unit, to.value, to.unit, options))
    {
      if (options != null)
        options.track(name, from.value, from.unit, to.value, to.unit);
      equals=false;
    }
    if(options != null)
      options.trackToMap();
    return equals;
  }
  
  public static boolean equals(double fValue, double tValue)
  {
    return equals(fValue, tValue, null);
  }
  // Tolerance is a percentage, not a decimal value
  public static boolean equals(double fValue, double tValue, DoubleUtils opts)
  {
    return equals(fValue,"",tValue,"", opts);
  }
  public static boolean equals(double fValue, String fUnit, double tValue, String tUnit)
  {
    return equals(fValue,fUnit,tValue,tUnit,null);
  }
  public static boolean equals(double fValue, String fUnit, double tValue, String tUnit, DoubleUtils options)
  {
    if (!DoubleUtils.equals(fValue, fUnit, tValue, tUnit, options))
      return false;
    return true;
  }
  public int hashCode()
  {
    int result = 1;

    long bits = Double.doubleToLongBits(this.value);
    result += this.unit.hashCode();
    result += (int)(bits^(bits >>> 32));

    return result;
  }
  
  /**
   * Set the amount of precision when converting to a string
   * @param p
   * @return
   */
  public void setPrecision(int p)
  {
  	this.precision=p;
  }
  
  public int getPrecision()
  {
  	return this.precision;
  }
  
  public void truncate(int mantissaLength)
  {
    this.value=DoubleUtils.truncate(this.value,mantissaLength);
  }
  
  public boolean isValid()
  {
  	if(Double.isNaN(this.value))
  		return false;
    if (!this.validUnit(this.unit))
      return false;
    return true;
  }

  public void invalidate()
  {
    this.unit="";
    this.value=Double.NaN;
    this.precision=-1;
  }

  public boolean validUnit(String unit)
  {
    return true;
  }

  public String getUnit()
  {
    return this.unit;
  }
  
  public double getValue()
  {
    if(Double.isNaN(this.value))
      Log.error("SEScalar Returning NaN from getValue()");
    return this.value;
  }

  public double getValue(String unit)
  {
    if(unit==null)
      return getValue();
    // Can't translate what we don't have
    if(Double.isNaN(this.value))
      return this.value;
    if (unit.equals(this.unit))
      return this.value;

    if (!validUnit(unit))
    {
      Log.error("Returning NaN");
      return Double.NaN;
    }
    return UnitConverter.convert(value, this.unit, unit);
  }

  public void setValue(double val)
  {
  	if(this.validUnit(""))
    {
      this.unit = "";
      this.value = val;
    }
    else
    {
      this.unit = "";
      this.value = Double.NaN;
      Log.error("Unit not provided on derived scalar type "+this.getClass());
      throw new InvalidUnitException(unit,this.getClass().getName());
    }
  }
  
  public void setValue(double val, int precision)
  {
    this.precision=precision;
    this.setValue(val);
  }

  public void setValue(double val, String unit)
  {
    if (Double.isNaN(val) || validUnit(unit))
    {// Don't care what the unit is if we are NaN, it's probably not good anyway
      if(unit==null)// NaN check is from if you request an unsupported compartment, you don't get a unit back
        this.unit="";
      else
        this.unit = unit;
      this.value = val;
    }
    else 
    {
      this.unit="";
      this.value=Double.NaN;
      throw new InvalidUnitException(unit,this.getClass().getName());
    }
  }
  
  public void setValue(double val, String unit, int precision)
  {
    this.precision=precision;
    this.setValue(val,unit);
  }
  
  public String toString()
  {
  	return this.value+this.unit;
  }
  
  public static String toString(ScalarData data)
  {
    return data.getValue()+data.getUnit();
  }
  
  public static SEScalar getScalar(Object object, String name)
  {    
    try
    {
      for(Method m : object.getClass().getMethods())
      {
        if(m.getName().startsWith("get") && m.getParameterCount()==0)
        {
          if(SEScalar.class.isAssignableFrom(m.getReturnType()))
          {
            if(m.getName().substring(3).equals(name))
              return (SEScalar)m.invoke(object);
          }
          else if(m.getReturnType().getSimpleName().startsWith("SE"))
          {
            SEScalar s = getScalar(m.invoke(object),name);
            if(s!=null)
              return s;
          }
        }
      }
    }
    catch (Exception ex)
    {
      
    }
    return null;
  }
}

File: projects/biogears/libBiogears/src/cdm/properties/SEScalar.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

#include <biogears/cdm/properties/SEScalar.h>

#include <cassert>
#include <cmath>
#include <limits>
#include <sstream>

#include "io/cdm/Property.h"

#include <biogears/cdm/utils/GeneralMath.h>

namespace biogears {
double SEScalar::NaN = std::numeric_limits<double>::quiet_NaN();

double SEScalar::dNaN()
{
  return NaN;
}
//-------------------------------------------------------------------------------
const NoUnit NoUnit::unitless = NoUnit();

const std::string unitless = "unitless";

NoUnit::NoUnit()
{
}
//-------------------------------------------------------------------------------
NoUnit::~NoUnit()
{
}
//-------------------------------------------------------------------------------
SEScalar::SEScalar(double v, bool ro)
  : SEScalar()

{
  SetValue(v);
}
//-------------------------------------------------------------------------------
SEScalar::SEScalar()
  : SEProperty()
  , m_readOnly(false)
  , m_value(1.0)
{
  Clear();
}
//-------------------------------------------------------------------------------
SEScalar::SEScalar(SEScalar const& obj) 
: SEProperty( obj)
  , m_readOnly(obj.m_readOnly)
  , m_value(obj.m_value)
{

}
//-------------------------------------------------------------------------------
SEScalar::SEScalar(SEScalar&& obj) 
: SEProperty()
  , m_readOnly(std::exchange(obj.m_readOnly, false))
  , m_value(std::exchange(obj.m_value, 1.0))
{

}
//-------------------------------------------------------------------------------
SEScalar::~SEScalar()
{
}

//-------------------------------------------------------------------------------
void SEScalar::Clear()
{
  m_readOnly = false;
  Invalidate();
}

//-------------------------------------------------------------------------------
bool SEScalar::Set(const SEScalar& s)
{
  if (!s.IsValid()) {
    return false;
  }
  if (m_readOnly) {
#if defined(BIOGEARS_THROW_READONLY_EXCEPTIONS)
    throw CommonDataModelException("Scalar is marked read-only");
#else
    return false;
#endif
  }
  m_value = s.m_value;
  return true;
}

//-------------------------------------------------------------------------------
void SEScalar::Copy(const SEScalar& s)
{
  if (m_readOnly) {
#if defined(BIOGEARS_THROW_READONLY_EXCEPTIONS)
    throw CommonDataModelException("Scalar: " + s.ToString() + "is marked read-only");
#else
    return;
#endif
  }
  m_value = s.m_value;
}

//-------------------------------------------------------------------------------
void SEScalar::Invalidate()
{
  if (m_readOnly) {
#if defined(BIOGEARS_THROW_READONLY_EXCEPTIONS)
    throw CommonDataModelException("Scalar is marked read-only");
#else
    return;
#endif
  }
  m_value = NaN;
}
#pragma optimize("", off)
//-------------------------------------------------------------------------------
bool SEScalar::IsValid() const
{
  return !std::isnan(m_value);
}
#pragma optimize("", on)
//-------------------------------------------------------------------------------
bool SEScalar::IsZero(double limit) const
{
  if (!IsValid()) {
    return false;
  }
  return SEScalar::IsZero(m_value, limit);
}

//-------------------------------------------------------------------------------
bool SEScalar::IsPositive() const
{
  if (!IsValid()) {
    return false;
  }
  return m_value > 0;
}

//-------------------------------------------------------------------------------
bool SEScalar::IsNegative() const
{
  if (!IsValid()) {
    return false;
  }
  return m_value < 0;
}

//-------------------------------------------------------------------------------
void SEScalar::SetReadOnly(bool b)
{
  m_readOnly = b;
}
//-------------------------------------------------------------------------------
bool SEScalar::IsReadOnly() const
{
  return m_readOnly;
}

//-------------------------------------------------------------------------------
double SEScalar::GetValue() const
{
#if defined(BIOGEARS_THROW_NAN_EXCEPTIONS)
  if (!IsValid()) {
    throw CommonDataModelException("Value is NaN");
  }
#endif
  return m_value;
}

//-------------------------------------------------------------------------------
void SEScalar::SetValue(double d)
{
  if (m_readOnly) {
#if defined(BIOGEARS_THROW_READONLY_EXCEPTIONS)
    throw CommonDataModelException("Scalar is marked read-only");
#else
    return;
#endif
  }
  m_value = d;
}

//-------------------------------------------------------------------------------
double SEScalar::Increment(const SEScalar& s)
{
  if (!s.IsValid())
    Invalidate();
  else
    IncrementValue(s.GetValue());
  return m_value;
}

//-------------------------------------------------------------------------------
double SEScalar::IncrementValue(double d)
{
  if (!IsValid()) {
    SetValue(d);
    return d;
  }
  SetValue(m_value + d);
  return m_value;
}
//-------------------------------------------------------------------------------
double SEScalar::Decrement(const SEScalar& s)
{
  if (!s.IsValid())
    Invalidate();
  else
    IncrementValue(-1.0 * s.GetValue());
  return m_value;
}
//-------------------------------------------------------------------------------
double SEScalar::DecrementValue(double d)
{
  return IncrementValue(-1.0 * d);
}
//-------------------------------------------------------------------------------
double SEScalar::Multiply(const SEScalar& s)
{
  if (!s.IsValid())
    Invalidate();
  else
    MultiplyValue(s.GetValue());
  return m_value;
}
//-------------------------------------------------------------------------------
double SEScalar::MultiplyValue(double d)
{
  if (!IsValid()) {
    SetValue(d);
    return d;
  }
  SetValue(m_value * d);
  return m_value;
}
//-------------------------------------------------------------------------------
double SEScalar::Divide(const SEScalar& s)
{
  if (!s.IsValid())
    Invalidate();
  else
    DivideValue(s.GetValue());
  return m_value;
}
//-------------------------------------------------------------------------------
double SEScalar::DivideValue(double d)
{
  if (!IsValid()) {
    SetValue(1.0 / d);
    return 1 / d;
  }
  SetValue(m_value / d);
  return m_value;
}
//-------------------------------------------------------------------------------
bool SEScalar::Equals(const SEScalar& to) const
{
  if (std::isnan(m_value) && std::isnan(to.m_value)) // This Violates C++ Spec
    return true;
  if (std::isnan(m_value) || std::isnan(to.m_value))
    return false;
  if (std::isinf(m_value) && std::isinf(to.m_value)) // This implies -> -inf == +inf
    return true;
  if (std::isinf(m_value) || std::isinf(to.m_value))
    return false;
  return std::abs(GeneralMath::PercentDifference(m_value, to.m_value)) <= 1e-15;
}

//-------------------------------------------------------------------------------
std::string SEScalar::ToString() const
{
#ifndef ANDROID
  return std::to_string(m_value);
#else
  std::stringstream ss;
  ss << m_value;
  return ss.str();
#endif
}

//-------------------------------------------------------------------------------
bool SEScalar::IsValue(double target, double value)
{
  if (value < (target + 1e-10) && value > (target - 1e-10))
    return true;
  return false;
}

//-------------------------------------------------------------------------------
bool SEScalar::IsZero(double d, double limit)
{
  if (d < limit && d > -limit)
    return true;
  return false;
}
//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator=(const SEScalar& rhs)
{
  if (this == &rhs)
    return *this;
  auto temp = SEScalar(rhs);
  SEProperty::operator=(temp);
  this->m_value = rhs.m_value;
  this->m_readOnly = rhs.m_readOnly;

  return *this;
}

//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator=(SEScalar&& rhs)
{
  if (this == &rhs)
    return *this;
  auto temp = SEScalar(std::move(rhs));
  SEProperty::operator=(std::move(temp));
  std::swap(m_value, temp.m_value);
  std::swap(m_readOnly, temp.m_readOnly);
  return *this;
}
//-------------------------------------------------------------------------------
bool SEScalar::operator<(const SEScalar& rhs) const
{
  if (!IsValid() || !rhs.IsValid()) {
    return false;
  }
  return m_value < rhs.m_value;
}
//-------------------------------------------------------------------------------
bool SEScalar::operator<=(const SEScalar& rhs) const
{
  if (!IsValid() || !rhs.IsValid()) {
    return false;
  }
  return m_value <= rhs.m_value;
}
//-------------------------------------------------------------------------------
bool SEScalar::operator>(const SEScalar& rhs) const
{
  if (!IsValid() || !rhs.IsValid()) {
    return false;
  }
  return m_value > rhs.m_value;
}
//-------------------------------------------------------------------------------
bool SEScalar::operator>=(const SEScalar& rhs) const
{
  if (!IsValid() || !rhs.IsValid()) {
    return false;
  }
  return m_value >= rhs.m_value;
}
//-------------------------------------------------------------------------------
SEScalar SEScalar::operator+(const SEScalar& rhs) const
{

  SEScalar result;
  result.Set(*this);
  result.Increment(rhs);
  return result;
}
//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator+=(const SEScalar& rhs)
{
  Increment(rhs);
  return *this;
}
//-------------------------------------------------------------------------------
SEScalar SEScalar::operator-(const SEScalar& rhs) const
{
  SEScalar result;
  result.Set(*this);
  result.Decrement(rhs);
  return result;
}
//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator-=(const SEScalar& rhs)
{
  Decrement(rhs);
  return *this;
}
//-------------------------------------------------------------------------------
SEScalar SEScalar::operator/(const SEScalar& rhs) const
{
  SEScalar result;
  result.Set(*this);
  result.Divide(rhs);
  return result;
}
//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator/=(const SEScalar& rhs)
{
  Divide(rhs);
  return *this;
}
//-------------------------------------------------------------------------------
SEScalar SEScalar::operator*(const SEScalar& rhs) const
{
  SEScalar result;
  result.Set(*this);
  result.Multiply(rhs);
  return result;
}
//-------------------------------------------------------------------------------
SEScalar& SEScalar::operator*=(const SEScalar& rhs)
{
  Multiply(rhs);
  return *this;
}
} // namespace Biogears